repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/contract/AbstractContractPreProcessor.java

```diff
@@ -0,0 +1,104 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.apache.atlas.AtlasErrorCode.INSTANCE_BY_UNIQUE_ATTRIBUTE_NOT_FOUND;
+import static org.apache.atlas.AtlasErrorCode.TYPE_NAME_INVALID;
+import static org.apache.atlas.repository.Constants.*;
+
+public abstract class AbstractContractPreProcessor implements PreProcessor {
+    private static final Logger LOG = LoggerFactory.getLogger(AbstractContractPreProcessor.class);
+
+    public final AtlasTypeRegistry typeRegistry;
+    public final EntityGraphRetriever entityRetriever;
+    public final AtlasGraph graph;
+
+
+    AbstractContractPreProcessor(AtlasGraph graph, AtlasTypeRegistry typeRegistry,
+                                 EntityGraphRetriever entityRetriever) {
+        this.graph = graph;
+        this.typeRegistry = typeRegistry;
+        this.entityRetriever = entityRetriever;
+    }
+
+    void authorizeContractCreateOrUpdate(AtlasEntity contractEntity, AtlasEntity.AtlasEntityWithExtInfo associatedAsset) throws AtlasBaseException {
+        AtlasPerfMetrics.MetricRecorder metricRecorder = RequestContext.get().startMetricRecord("authorizeContractUpdate");
+        try {
+            AtlasEntityHeader entityHeader = new AtlasEntityHeader(associatedAsset.getEntity());
+
+            //First authorize entity update access
+            verifyAssetAccess(entityHeader, AtlasPrivilege.ENTITY_UPDATE, contractEntity, AtlasPrivilege.ENTITY_UPDATE);
+
+        } finally {
+            RequestContext.get().endMetricRecord(metricRecorder);
+        }
+    }
+
+
+    private void verifyAssetAccess(AtlasEntityHeader asset, AtlasPrivilege assetPrivilege,
+                                   AtlasEntity contract, AtlasPrivilege contractPrivilege) throws AtlasBaseException {
+        verifyAccess(asset, assetPrivilege);
+        verifyAccess(contract, contractPrivilege);
+    }
+
+    private void verifyAccess(AtlasEntity entity, AtlasPrivilege privilege) throws AtlasBaseException {
+        verifyAccess(new AtlasEntityHeader(entity), privilege);
+    }
+
+    private void verifyAccess(AtlasEntityHeader entityHeader, AtlasPrivilege privilege) throws AtlasBaseException {
+        String errorMessage = privilege.name() + " entity: " + entityHeader.getTypeName();
+        AtlasAuthorizationUtils.verifyAccess(new AtlasEntityAccessRequest(typeRegistry, privilege, entityHeader), errorMessage);
+    }
+
+    AtlasEntity.AtlasEntityWithExtInfo getAssociatedAsset(String datasetQName, String typeName) throws AtlasBaseException {
+
+        Map<String, Object> uniqAttributes = new HashMap<>();
+        uniqAttributes.put(QUALIFIED_NAME, datasetQName);
+
+        AtlasEntityType entityType = ensureEntityType(typeName);
+
+        AtlasVertex entityVertex = AtlasGraphUtilsV2.getVertexByUniqueAttributes(graph, entityType, uniqAttributes);
+
+        EntityGraphRetriever entityRetriever = new EntityGraphRetriever(graph, typeRegistry, true);
```

Comment:
I believe injecting should be better instead of creating a new instance (I will remove creating new object). As it will pass the reference and hence reduce the unnecessary memory utilisation. Let me know your thoughts.

Reply:
agree